pkg/cloudprovider/instances.go

```diff
@@ -0,0 +1,88 @@
+package cloudprovider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+const (
+	InternalIPLabel = "k3s.io/internal-ip"
+	ExternalIPLabel = "k3s.io/external-ip"
+	HostnameLabel   = "k3s.io/hostname"
+)
+
+func (k *k3s) AddSSHKeyToAllInstances(ctx context.Context, user string, keyData []byte) error {
+	return cloudprovider.NotImplemented
+}
+
+func (k *k3s) CurrentNodeName(ctx context.Context, hostname string) (types.NodeName, error) {
+	return types.NodeName(hostname), nil
+}
+
+func (k *k3s) InstanceExistsByProviderID(ctx context.Context, providerID string) (bool, error) {
+	return true, cloudprovider.NotImplemented
+}
+
+func (k *k3s) InstanceID(ctx context.Context, nodeName types.NodeName) (string, error) {
+	_, err := k.NodeClient.Get(string(nodeName), metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("Failed to find node %s: %v", nodeName, err)
+	}
+	return string(nodeName), nil
+}
+
+func (k *k3s) InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error) {
+	return true, cloudprovider.NotImplemented
+}
+
+func (k *k3s) InstanceType(ctx context.Context, name types.NodeName) (string, error) {
+	_, err := k.InstanceID(ctx, name)
+	if err != nil {
+		return "", err
+	}
+	return "k3s", nil
+}
+
+func (k *k3s) InstanceTypeByProviderID(ctx context.Context, providerID string) (string, error) {
+	return "", cloudprovider.NotImplemented
+}
+
+func (k *k3s) NodeAddresses(ctx context.Context, name types.NodeName) ([]corev1.NodeAddress, error) {
+	addresses := []corev1.NodeAddress{}
+	node, err := k.NodeClient.Get(string(name), metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find node %s: %v", name, err)
+	}
+	// check internal address
+	if addr, ok := node.Labels[InternalIPLabel]; ok || addr != "" {
```

Comment:
Do you want to allow empty strings?  I think the logic you want here is just `if node.Labels[InternalIPLabel] != ""`